psp34_shiden/contracts/shiden34/src/lib.rs

```diff
@@ -0,0 +1,381 @@
+            if Self::env().transferred_value() != mint_amount as u128 * self.price_per_mint {
+                return Err(PSP34Error::Custom("BadMintValue".to_string()))
+            }
+            Ok(())
+        }
+
+        /// Check amount of tokens to be minted
+        fn check_amount(&self, mint_amount: u64) -> Result<(), PSP34Error> {
+            if mint_amount == 0 {
+                return Err(PSP34Error::Custom("CannotMintZeroTokens".to_string()))
+            }
+
+            if self.last_token_id + mint_amount > self.max_supply {
+                return Err(PSP34Error::Custom("CollectionFullOrLocked".to_string()))
+            }
+            Ok(())
+        }
+
+        /// Check if token is minted
+        fn token_exists(&self, id: Id) -> Result<(), PSP34Error> {
+            self.owner_of(id).ok_or(PSP34Error::TokenNotExists)?;
+            Ok(())
+        }
+    }
+
+    impl PSP34Mintable for Shiden34Contract {
+        /// This mint function is not used since it can't be made payable
+        #[ink(message)]
+        fn mint(&mut self, _: AccountId, _: Id) -> Result<(), PSP34Error> {
+            Ok(())
+        }
+    }
+
+    impl Shiden34Trait for Shiden34Contract {
+        /// Mint next available token for the caller
+        #[ink(message, payable)]
+        fn mint_next(&mut self) -> Result<(), PSP34Error> {
+            self.check_value(1)?;
+            let caller = self.env().caller();
+            self.last_token_id += 1;
+            self._mint_to(caller, Id::U64(self.last_token_id))?;
+
+            Ok(())
+        }
+
+        /// Mint several tokens
+        #[ink(message, payable)]
+        #[modifiers(non_reentrant)]
+        fn mint_for(&mut self, to: AccountId, mint_amount: u64) -> Result<(), PSP34Error> {
+            self.check_value(mint_amount)?;
+            self.check_amount(mint_amount)?;
+
+            let next_to_mint = self.last_token_id + 1; // first mint id is 1
+            let mint_offset = next_to_mint + mint_amount;
+
+            for mint_id in next_to_mint..mint_offset {
+                assert!(self._mint_to(to, Id::U64(mint_id)).is_ok());
+                self.last_token_id += 1;
+            }
+
+            Ok(())
+        }
+
+        /// Set new value for the baseUri
+        #[ink(message)]
+        #[modifiers(only_owner)]
+        fn set_base_uri(&mut self, uri: String) -> Result<(), PSP34Error> {
+            self._set_attribute(
+                Id::U8(0),
+                String::from("baseUri").into_bytes(),
+                uri.into_bytes(),
+            );
+            Ok(())
+        }
+
+        /// Get URI from token ID
+        #[ink(message)]
+        fn token_uri(&self, token_id: u64) -> Result<String, PSP34Error> {
+            _ = self.token_exists(Id::U64(token_id))?;
+            let value = self.get_attribute(Id::U8(0), String::from("baseUri").into_bytes());
+            let mut token_uri = String::from_utf8(value.unwrap()).unwrap();
```

Comment:
unwrap() could panic!

Reply:
I guess no harm in panic since this is contract, and this call is read call